osutil/syncdir.go

```diff
@@ -22,18 +22,62 @@ package osutil
 import (
 	"bytes"
 	"fmt"
+	"io"
 	"io/ioutil"
 	"os"
 	"path/filepath"
 	"sort"
 )
 
-// FileState describes the expected content and meta data of a single file.
-type FileState struct {
+// FileState is an interface for conveying the desired state of a some file.
+type FileState interface {
+	State() (reader io.ReadCloser, size int64, mode os.FileMode, err error)
+}
+
+// FileReference describes the desired content by referencing an existing file.
+type FileReference struct {
+	Path string
+}
+
+// State returns a reader of the referenced file, along with other meta-data.
+func (fref FileReference) State() (io.ReadCloser, int64, os.FileMode, error) {
+	file, err := os.Open(fref.Path)
+	if err != nil {
+		return nil, 0, os.FileMode(0), err
+	}
+	fi, err := file.Stat()
+	if err != nil {
+		return nil, 0, os.FileMode(0), err
+	}
+	return file, fi.Size(), fi.Mode(), nil
+}
+
+// FileContentReference describes the desired content by referencing an existing file and providing custom mode.
+type FileContentReference struct {
```

Comment:
not sure why there is "Content" in the name of this one?  all of these are about file content. Suggestion:

FileReferencePlusMode

Reply:
I used `content` because the code iterating over various maps was using `content` as the variable but I'm not attached to it. I applied the suggested name.

Reply:
Renamed.